pkg/crds/crds.go

```diff
@@ -0,0 +1,192 @@
+	//go:embed calico
+	calicoCRDFiles embed.FS
+	//go:embed enterprise
+	enterpriseCRDFiles embed.FS
+	//go:embed operator/*
+	operatorCRDFiles embed.FS
+
+	yamlDelimRe       *regexp.Regexp
+	calicoOprtrCRDsRe *regexp.Regexp
+
+	// We cache these CRDs because to generate the calico and enterprise takes
+	// approximately 40ms, with the caching 1ms.
+	lock           sync.Mutex
+	calicoCRDs     []*apiextenv1.CustomResourceDefinition
+	enterpriseCRDs []*apiextenv1.CustomResourceDefinition
+)
+
+func init() {
+	yamlDelimRe = regexp.MustCompile(`\n---`)
+
+	calicoCRDNames := []string{"installation", "apiserver", "imageset", "tigerastatus"}
+	calicoOprtrCRDsRe = regexp.MustCompile(fmt.Sprintf("(%s)", strings.Join(calicoCRDNames, "|")))
+}
+
+func getCalicoCRDSource() map[string][]byte {
+	ret := map[string][]byte{}
+	entries, err := calicoCRDFiles.ReadDir("calico")
+	if err != nil {
+		panic(fmt.Sprintf("Failed to read Calico CRDs: %v", err))
+	}
+
+	for _, entry := range entries {
+		b, err := calicoCRDFiles.ReadFile(path.Join("calico", entry.Name()))
+		if err != nil {
+			panic(fmt.Sprintf("Failed to read Calico CRD %s: %v", entry.Name(), err))
+		}
+
+		if len(yamlDelimRe.FindAllString(string(b), -1)) > 1 {
+			panic(fmt.Sprintf("Too many yaml delimiters in Calico CRD %s", entry.Name()))
+		}
+
+		ret[entry.Name()] = yamlDelimRe.ReplaceAll(b, []byte("\n"))
+	}
+
+	return ret
+}
+
+func getEnterpriseCRDSource() map[string][]byte {
+	ret := map[string][]byte{}
+	entries, err := enterpriseCRDFiles.ReadDir("enterprise")
+	if err != nil {
+		panic(fmt.Sprintf("Failed to read Enterprise CRDs: %v", err))
+	}
+
+	for _, entry := range entries {
+		b, err := enterpriseCRDFiles.ReadFile(path.Join("enterprise", entry.Name()))
+		if err != nil {
+			panic(fmt.Sprintf("Failed to read Enterprise CRD %s: %v", entry.Name(), err))
+		}
+
+		if len(yamlDelimRe.FindAllString(string(b), -1)) > 1 {
+			panic(fmt.Sprintf("Too many yaml delimiters in Enterprise CRD %s", entry.Name()))
+		}
+
+		ret[entry.Name()] = yamlDelimRe.ReplaceAll(b, []byte("\n"))
+	}
+
+	return ret
+}
+
+func getOperatorCRDSource(variant opv1.ProductVariant) map[string][]byte {
+	ret := map[string][]byte{}
+	entries, err := operatorCRDFiles.ReadDir("operator")
+	if err != nil {
+		panic(fmt.Sprintf("Failed to read Operator CRDs: %v", err))
+	}
+
+	for _, entry := range entries {
+		if variant == opv1.Calico {
+			if !calicoOprtrCRDsRe.MatchString(entry.Name()) {
+				continue
+			}
+		}
+
+		b, err := operatorCRDFiles.ReadFile(path.Join("operator", entry.Name()))
+		if err != nil {
+			panic(fmt.Sprintf("Failed to read Operator CRD %s: %v", entry.Name(), err))
+		}
+
+		if len(yamlDelimRe.FindAllString(string(b), -1)) > 1 {
+			panic(fmt.Sprintf("Too many yaml delimiters in Operator CRD %s", entry.Name()))
+		}
+
+		ret[entry.Name()] = yamlDelimRe.ReplaceAll(b, []byte("\n"))
+	}
+
+	return ret
+}
+
+func GetCRDs(variant opv1.ProductVariant) []*apiextenv1.CustomResourceDefinition {
+	lock.Lock()
```

Comment:
Why do you need a lock? What's the race you're trying to prevent?

Reply:
I'm trying to prevent if there were multiple calls happening writing the global calicoCRDs or enterpriseCRDs.

Reply:
Is there a write here that I missed?

Reply:
This is only reading from the files and returning the results, or am I missing something? Two simultaneous calls to this would conflict.

Reply:
Where the CRDs are being stored is global. So by write I meant appending to calicoCRDs or enterpriseCRDs.

Reply:
OK, this is what I was missing, I didn't realise that `calicoCRDs` and `enterpriseCRDs` were global. The lock makes sense now to me now. Thanks for explaining.